### PR TITLE
add support for MBEDTLS_SSL_VERIFY_EXTERNAL authmode

### PR DESCRIPTION
**Pull request by awakecoding**
This pull request adds a new "external" certificate verification mode that makes it possible for an application to override the certificate validation logic with much more flexibility and simplicity. I took special care in adding the new mode in a way that would make it easy to detect at compilation time and would not cause an API or ABI breakage.

Here is the list of changes:

1. New MBEDTLS_SSL_VERIFY_EXTERNAL (4) definition, can be used for compile-time detection
2. unsigned int autmode : 2 was changed to : 3 because the value 4 did not fit in 2 bits.
3. When the external mode is used, the verification callback is called with the certificate chain sent by the server and a depth of -1 to indicate that the entire chain, not a single certificate, is passed.

I have set the depth to -1 because in this case it should not mean anything, since we're passing the chain and not one certificate at a time like what I understood happens with some other modes.

The rationale for exposing the certificate chain sent by the server more directly to the application is that some cases cannot be easily covered without "polluting" the core library with too much application-specific or platform-specific logic. In the simplest case, one can just call the mbedtls APIs again with what has been passed in the callback to validate the certificate chain, and then handle validation failure in a custom way.

I see two major use cases for external certificate validation:

1) Application-specific handling of certificate exceptions. If automatic validation fails, the application can then present the certificate chain to the user for manual validation. The application may also save a hash of certificates that were marked as exceptions by the user, and add the exception list to the validation logic. This does not prevent the application by trying to validate the certificate chain with the same mbed TLS APIs used for automatic certificate validation, but it makes it easy to insert additional logic without touching the mbed TLS core libraries.

2) Validation against platform-specific certificate stores, the most notable one being the Windows certificate store that is not exposed in the form of a directory with PEM files. The mbed TLS API works well with file-based certificate stores and certificate bundles, but it does not support platform-specific certificate APIs, and as far as I know, it is not the goal of the project to do so. By exposing the certificate chain more directly to the application, it becomes easier to call platform-specific APIs to perform the validation against the system trusted certificates.

I have an mbed account with the same email used for the commit in this PR and I have accepted the contribution agreement. Please let me know what you think, I am open to making changes but I really think mbed TLS needs to allow external certificate validation logic in one way or another.

### CODE REVIEW

**Comment by hanno-arm** [comment]
Hi @awakecoding,

thank you very much for your contribution, and in particular for providing a detailed PR description.

I would like to discuss the rationale for this feature enhancement in more detail before considering the proposed code changes. Specifically, what part of 1. and 2. cannot be done using the existing verification callback mechanism? With the latter, the user has full control over removing or adding verification failure flags after Mbed TLS has attempted the verification; in particular, it allows users to add exceptions by clearing _specific_ verification failure flags and hence having the verification succeed provided there were no other problems. If the entire verification mechanism is offloaded, one is at the risk of waving through a certificate which has multiple problems just because one of these problems is known and tolerated.

The use case 2. is indeed a known limitation of Mbed TLS: we currently don't support _direct_ registration of external trusted CA storages. However, there is a workaround with the present verification callback mechanism, too: You omit the configuration of trusted CAs through `mbedtls_ssl_conf_ca_chain()`, and instead register a verification callback which checks for certificates which failed with `MBEDTLS_X509_BADCERT_NOT_TRUSTED`: those for which no parent was found, neither in the rest of the chain nor in the trusted CA pool. When it finds such, it searches the external trusted CA storage   to see if the parent is trusted, verifies the signature, and clears `MBEDTLS_X509_BADCERT_NOT_TRUSTED` on success. This will make the verification succeed if there were no other problems with the chain.

Let me emphasize that this is not a rejection of your contribution, but an attempt to better understand the rationale for it first, and to check that its use case isn't sufficiently covered by the existing API.

cc @mpg who wrote most of the X.509 verification logic.

Thank you again,
Kind regards, 
Hanno, Mbed TLS team member

**Comment by RonEld** [comment]
@awakecoding As my colleague mentioned, we thank you for your contribution.

In addition, unfortunately our policy is to not accept contributions, without a Contributor’s Licence Agreement (CLA) signed or authorised by yourself or your employer.

If this is a personal contribution, the easiest way to do this is if you create an mbed account and [accept this click through agreement](https://os.mbed.com/contributor_agreement/). Alternatively, you can find a [slightly different agreement to sign here](https://www.mbed.com/en/about-mbed/contributor-license-agreements/), which can be signed and returned to us, and is applicable if you don't want to create an mbed account or alternatively if this is a corporate contribution.

Thanks for your understanding and again, thanks for the contribution!

**Comment by awakecoding** [comment]
@RonEld sorry about that, I missed the part where I also needed a signed CLA from my employer. I am aware of the implications of such an agreement but this will be a first in this company, I will see that it gets done as soon as possible.

**Comment by RonEld** [comment]
@awakecoding Thank you for your cooperation. Are you contributing on behalf of your company? If so, you will need a corporate CLA signed by someone authorized to sign for your company.
Thank you in advance!

**Comment by awakecoding** [comment]
@RonEld I have just sent the signed CLA for Devolutions (I am authorized to sign) to cla-submissions@arm.com. Please let me know when you will have had the time to review that everything is in order and that I am good to go to make contributions :)

**Comment by awakecoding** [comment]
@hanno-arm @mpg thank you for the detailed response.

To explain where I am coming from with this, I should mention that we are using mbed TLS in a peer-to-peer remote desktop product called [Wayk Now](http://wayk.devolutions.net/). I do not use mbed TLS in "real embedded systems", I use it on the major desktop and mobile platforms (Windows, macOS, Linux, Android and iOS). When direct connections are made, a self-signed certificate is used, meaning that certification validation will almost certainly fail. We do have automated certificate validation when connections are made through our central server as described in this [blog post](https://blog.devolutions.net/2018/09/insider-series-wayk-den-certificate-authority).

Here are my comments regarding both use cases and the current limitations of the API:

1) The way I understood it, the current API can pass each individual certificate of the chain to the application through a callback, and allow clearing specific validation flags. This is not such a bad thing, but it is impractical if you want to get the entire certificate chain at once to present it to the user when certificate validation fails. Since the current API calls the callback once per certificate, the application needs to manually rebuild the chain over multiple calls, unless I am mistaken.

The way I see it, I would rather get the complete certificate chain and pass it to the mbed TLS certificate validation API to get the same result that is currently obtained under the hood. The difference is the following: the application can have complete control over the certificate validation logic, instead of being called by mbed TLS as part of its validation logic. This makes much easier to introduce application-specific extensions, such as saving a hash of a certificate that the user manually accepted, in a way that is very similar to SSH's known_hosts file. When certificate validation fails in Wayk Now, we present the entire certificate chain to the user and ask for manual validation.

The idea is not to skip the same validation calls already made under regular conditions, it is to have the same call made from an application function that has more control over the complete validation process. The information returned by the mbed TLS validation functions is very useful, but it is much easier to call it from the application instead of having the validation function call the application through more fine grained callbacks as it performs validation. I don't think one should prevent the other, and if you prefer, we may even add a separate callback to "externalize" the validation process.

2) I see from your answer that there exists a possible way to make it work with the current API, but I would rather make it much easier to offload the validation logic entirely to the application. I haven't experimented with the Windows certificate store API yet, but what I'd like to implement is a single callback function that can take the certificate chain and proper parameters, and perform validation against the Windows certificate store. The alternative I've been thinking about would be to make a function that dumps the entire trusted root CAs in a file and load it every time, but that would be extremely wasteful. The main issue I see when using mbed TLS instead of the system-provided TLS stack is that we instantly loose the ability to perform validation against the system-provided trusted root CAs, unless the system certificate store is file-based.

This being said, I see two potential ways to address this issue:

A) Use the approach provided in this PR, which requires very minimal changes and doesn't make an API or ABI breakage.

B) Expose a new API to explicitly offload certificate validation to the application, but provide a simple way to call the mbed TLS validation API again with the information provided in the new API callbacks.

Let me know what you think

**Comment by hanno-arm** [comment]
Hi @awakecoding,

thanks for your detailed feedback.

I think it is worth considering allowing a user to offload the entire CRT verification if he/she wants to use a _different_ X.509 stack for verification. Of course, ideally we'd like Mbed TLS' X.509 implementation to be used, but the door should be open for alternatives. On the client, the closest you get to this with the current API is to set the authentication mode to `MBEDTLS_SSL_VERIFY_NONE` and to manually query and verify the peer's CRT using `mbedtls_ssl_get_peer_crt()` after the handshake completed. On the server, it is currently not possible to request a client CRT without verifying it; the only option is to use `MBEDTLS_SSL_VERIFY_OPTIONAL` and to again inspect the peer's CRT using `mbedtls_ssl_get_peer_crt()` afterwards, but in this case, Mbed TLS will still have performed its own validation, and while it's ok to ignore the results, it's inefficient to run the verification in the first place if the users doesn't need it.

_However_, as far as I understand you, you _do_ want to make use of Mbed TLS' X.509 verification abilities, but are not satisfied with the way the API allows you to oversee and influence the verification process - correct? Before replying to your specific requests, I think in this situation it is still desirable to stick to the flow of (a) Mbed TLS verifies. (b) Application reviews and modifies results through callback. (c) Mbed TLS accepts/rejects on the basis of the changes of the application, for two reasons:
- Calling the verification logic without a callback and inspecting the failure flags afterwards gives less fine grained information on the nature of the failure, because the failure flags from the various CRTs in the chain have been merged. For example, if you get a `MBEDTLS_X509_BADCERT_EXPIRED` flag, you don't know to which CRT in the chain it applies. In contrast, the level-wise callbacks allow to make this distinction. You might want to white-list some failures for specific levels only, which is not possible without level-wise callbacks.
- Reiterating a point from my last reply, but it's worth emphasizing -- it's easy to wave through a CRT which is invalid for multiple reasons because one of them is known and tolerated.

But now to your concrete problems. First point:
>  ... impractical if you want to get the entire certificate chain at once to present it to the user when certificate validation fails. Since the current API calls the callback once per certificate, the application needs to manually rebuild the chain over multiple calls, unless I am mistaken.

While I don't see why one would want to allow users to tolerate CRT verification failure in your context, I agree that it's inconvenient that the callback doesn't have access to the entire chain + failures at once. At the time the callbacks were introduced, this was a limitation enforced by the way verification worked, but by now things have changed.

__Potential solution:__ If we changed the signature of the verification callback to pass the entire CRT chain as constructed by the verification stack, plus the failure flags for each level, hence allow you to inspect the entire chain at once while keeping fine-grained access to the depth-specific flags - would that solve your problems? 

Second point:

> I see from your answer that there exists a possible way to make it work with the current API, but I would rather make it much easier to offload the validation logic entirely to the application. 

How is it getting easier if you offload the validation logic to the application, _provided_ you want to continue using Mbed TLS' X.509 stack in the callback? You'd still face the limitation that the trusted CA's must currently be provided as a linked list of `mbedtls_x509_crt` instances - which, as far as I see, can currently only be overcome in the way described above, i.e. by not registering any CAs and manually searching for them after the verification.

In summary, I see three potential enhancements to Mbed TLS:
1. Offload X.509 verification entirely, for users who want to use a separate X.509 stack.
2. Modify verification callback to receive entire CRT chain and all failure flags at once, for users who need to white-list some known failure conditions and/or perform allow manual inspection.
3. Add a new callback for searching for trusted CAs, for users that need to use an external CA storage.

If I understood you correctly in that you'd like to stick to Mbed TLS' X.509 stack, I think (2) mainly allows to address your first problem, while (3) allows to address your second problem of using an external trusted CA storage. However, that second problem could also be solved through (2) with comparable effort.

Kind regards,
Hanno

**Comment by hanno-arm** [comment]
@awakecoding At this point I'd like to hear @mpg's opinion before proceeding. @mpg, could you have a look and provide your view on the matter?

**Comment by awakecoding** [comment]
@hanno-arm @mpg any update on this? this is a really important feature which I'd like to see upstream, since I'd like to integrate it in third-party libraries that also rely on mbedtls like nanomsg-next-generation.

**Comment by gdamore** [comment]
This seems like we have a first step here, and I'm curious about the progress as well -- these options may help my downstream users as well (I produce a communications library -- github.com/nanomsg/nng -- which utilizes mbedTLS.)

**Comment by hanno-arm** [comment]
Hi @awakecoding, 
my apologies for the late reply. Unfortunately, we are currently short on time and it is likely to take some more time until another team member can have a look at your contribution. I will try to remember the issue and look out for potential reviewers regularly, but please don't hold back to ping us in a week or two to see if things have changed.
Kind regards, 
Hanno

**Comment by sbutcher-arm** [comment]
@mpg - Can you please approve this for design? 

Labelling as 'needs design approval'.

**Comment by awakecoding** [comment]
@mpg @hanno-arm @sbutcher-arm any updates? I will update my PR to fix the conflict that was introduced since the last time I checked.

File: library/ssl_tls.c [hunk 1]
```diff
@@ -5744,13 +5744,24 @@ int mbedtls_ssl_parse_certificate( mbedtls_ssl_context *ssl )
...
                                 ssl->session_negotiate->peer_cert,
                                 ca_chain, ca_crl,
                                 ssl->conf->cert_profile,
                                 ssl->hostname,
                                &ssl->session_negotiate->verify_result,
                                 ssl->conf->f_vrfy, ssl->conf->p_vrfy, rs_ctx );
+        }
 
         if( ret != 0 )
         {
```
Comment (hanno-arm) [inline library/ssl_tls.c]: This debug message isn't accurate if the new authentication mode is used.

File: library/ssl_tls.c [hunk 2]
```diff
@@ -5744,13 +5744,24 @@ int mbedtls_ssl_parse_certificate( mbedtls_ssl_context *ssl )
         /*
          * Main check: verify certificate
          */
-        ret = mbedtls_x509_crt_verify_restartable(
+
+        if( authmode == MBEDTLS_SSL_VERIFY_EXTERNAL )
+        {
+            if( NULL != ssl->conf->f_vrfy )
```
Comment (hanno-arm) [inline library/ssl_tls.c]: Two points:
1. We might want to fail if `ssl->conf->f_vrfy` is unset, instead of silently succeeding.
2. The semantics of `f_vrfy` in the new authmode needs to be documented. The existing callbacks are expected to return `0` even if they find an issue with the CRT, and to instead the verification flags accordingly. In other words, all but non-fatal failures must be communicated through verification flags, but not through the return value. This is different with the suggested re-purposing of `f_vrfy` here, which - at least with the current code - would need to return a non-zero value for the verification to fail.

**Comment by hanno-arm** [comment]
@awakecoding I can only apologise again - due to other priorities, we haven't had the manpower to look into your proposal any further, and I don't know when it'll be the case. However, I left a few comments on your code - could you have a look, and also rebase the PR on top of current `development` to resolve merge conflicts? Thank you!

**Comment by mpg** [comment]
Hi @awakecoding  and sorry that we couldn't find the time to have a deeper look at your PR yet. However, regarding the second use case you mention:

> Validation against platform-specific certificate stores, the most notable one being the Windows certificate store that is not exposed in the form of a directory with PEM files. The mbed TLS API works well with file-based certificate stores and certificate bundles, but it does not support platform-specific certificate APIs, and as far as I know, it is not the goal of the project to do so. By exposing the certificate chain more directly to the application, it becomes easier to call platform-specific APIs to perform the validation against the system trusted certificates.

We've just merged a PR adding a callback that might help for that, so I wanted to let you know: https://github.com/ARMmbed/mbedtls/pull/2532

**Comment by awakecoding** [comment]
@mpg thanks, I have looked at https://github.com/ARMmbed/mbedtls/pull/2532 and it is definitely an improvement to the existing API. It looks like you guys are reworking the API related to certificate validation, in light of the recent changes, would the "external mode" suggested in this PR be implemented differently?

**Comment by JunqiWang** [comment]
> Offload X.509 verification entirely, for users who want to use a separate X.509 stack.

@awakecoding In case you're still interested in this. Would it be better if the callback passes byte array that carries list of certs received from server? In that case it is independent of mbedtls X509 completely. Or do you prefer mbedtls to parse the certs?

**Comment by awakecoding** [comment]
> > Offload X.509 verification entirely, for users who want to use a separate X.509 stack.
> 
> @awakecoding In case you're still interested in this. Would it be better if the callback passes byte array that carries list of certs received from server? In that case it is independent of mbedtls X509 completely. Or do you prefer mbedtls to parse the certs?

Passing the list of certificates as raw bytes would definitely make it easier to decouple from mbedtls for the validation. I would rather have that than some mbedtls types again, it wouldn't take much to reparse them with mbedtls again anyway.

It's been two years since I opened this, I still use the same modified validation callback, but now I have a lot of Rust code and at some point I will have to consider porting my validation code to Rust, so this would be useful.

**Comment by JunqiWang** [comment]
Thanks for the reply @awakecoding. FWIW I raised my ask in mbedtls mailing list and referenced this PR.

**Comment by mpg** [comment]
Hi, and sorry for not making the time to comment on this earlier. I've read all the past discussions and I fully agree with Hanno's summary

> In summary, I see three potential enhancements to Mbed TLS:
> 
>     1. Offload X.509 verification entirely, for users who want to use a separate X.509 stack.
> 
>     2. Modify verification callback to receive entire CRT chain and all failure flags at once, for users who need to white-list some known failure conditions and/or perform allow manual inspection.
> 
>     3. Add a new callback for searching for trusted CAs, for users that need to use an external CA storage.

I'd just add a few notes to this summary. Regarding (3), I think it was implemented in #2532 in the meantime.

Regarding (2):

- It's something I had in mind ever since introducing `mbedtls_x509_crt_verify_chain`: this would be a great structure to pass to the verify callback. There has been discussions above that the verify callback could in principle reconstruct the chain across calls, but that's at least extremely inconvenient (if not downright impossible) if you consider that a single `ssl_conf`, hence a single callback context, can be used for multiple connections (potentially concurrent).
- One potential drawback with exposing `mbedtls_x509_crt_verify_chain` is that it would make a currently-internal type part of the API. But I think the added flexibility for users would be well worth it. (We don't have to share exactly that structure, we could isolate a public type, and keep a private type which would potentially extend it.)
- One question that's been on my mind is whether this should be a new option, or completely replace the existing verify callback, as it's strictly more powerful. The main argument for replacing rather than adding would be to avoid having to maintain and test multiple interfaces. The main argument for adding is we can do it anytime, while replacing could only be done in 3.0 (which is already full) of 4.0 (which is likely a bit far in the future).

Regarding (1):

- I agree that it would be a valuable option to have, as it brings possibilities that are not covered by (2) and (3) at all.
- I agree that what we pass to the new callback should be bytes rather than using Mbed TLS types
- Ideally, with this option enabled, it would be possible to compile Mbed TLS without the X.509 module (in order to save code size) and still support some cert-based key exchanges in TLS (such as a client doing handshakes where only the server is authenticated) - however that would involve significantly more work than just creating this new authentication mode, so I'm just mentioning as an idea for the future, absolutely not at something that should be added to this PR.

Finally there's a difference between (1) and (2) that I'd like to highlight: with (1), the callback/application gets the chain _provided by the peer_ while with (2) it gets the chain _constructed by Mbed TLS_ (which is a subset of the peer's chain, followed by a trusted root). It think that point would be enough (even if there weren't already other reasons) to show that (1) and (2) would serve different purposes.

**Comment by JunqiWang** [comment]
Thanks @mpg for the summary. I'd vote for (1) since it covers what (2) and (3) can achieve, and unlocks other possibilities with the extended flexibility.
cc @daverodgman

**Comment by JunqiWang** [comment]
@awakecoding Would you like to update the PR to make the callback operate on the crt buffer rather than mbedtls_x509_crt type as suggested (if I understand @mpg's comment correctly😊 )?

**Comment by Rkcrl** [comment]
Pull request

**Comment by tom-cosgrove-arm** [comment]
Closing as no update to this PR in over a year.

If you want to fix the conflicts and rework based on mpg's suggestions, we would be happy to take another look.
